snakemake/sourcecache.py

```diff
@@ -126,13 +126,24 @@ def __init__(
         self.repo_path = repo_path
         self.path = path
 
+    def get_basedir(self):
+        root_path, file_path, ref = split_git_path(self.get_path_or_uri())
+        file_name = "{}/{}".format(root_path, file_path).replace(root_path, "")
+        return LocalGitFile("file:/{}".format(root_path), file_name, ref=ref)
+
     def get_path_or_uri(self):
         return "git+{}/{}@{}".format(self.repo_path, self.path, self.ref)
 
     def join(self, path):
+        if isinstance(path, LocalSourceFile):
+            path = path.get_path_or_uri()
```

Comment:
Seems cumbersome to me to have such special handling in a subclass like this. Other classes assume that path is a string. Why do you think this is needed?

Reply:
I tried to modify the code and ended up with having problems at

```
  File "/home/patsm159/workspace/snakemake/snakemake/utils.py", line 63, in validate
    schemafile = workflow.current_basedir.join(schemafile)
```

would it be ok to change 

`schemafile = workflow.current_basedir.join(schemafile)`

to

`schemafile = workflow.current_basedir.join(schemafile.get_path_or_uri())`